arch/arm/src/samv7/sam_qencoder.c

```diff
@@ -0,0 +1,378 @@
+  .inuse    = false,
+};
+#endif
+
+#ifdef CONFIG_SAMV7_TC2_QE
+static struct sam_lowerhalf_s g_tc2lower =
+{
+  .ops      = &g_qecallbacks,
+  .tcid     = 2,
+  .inuse    = false,
+};
+#endif
+
+#ifdef CONFIG_SAMV7_TC3_QE
+static struct sam_lowerhalf_s g_tc3lower =
+{
+  .ops      = &g_qecallbacks,
+  .tcid     = 3,
+  .inuse    = false,
+};
+#endif
+
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: sam_tc2lower
+ *
+ * Description:
+ *   Map a timer counter number to a device structure
+ *
+ ****************************************************************************/
+
+static FAR struct sam_lowerhalf_s *sam_tc2lower(int tc)
+{
+  switch (tc)
+    {
+#ifdef CONFIG_SAMV7_TC0_QE
+    case 0:
+      return &g_tc0lower;
+#endif
+#ifdef CONFIG_SAMV7_TC1_QE
+    case 1:
+      return &g_tc1lower;
+#endif
+#ifdef CONFIG_SAMV7_TC2_QE
+    case 2:
+      return &g_tc2lower;
+#endif
+#ifdef CONFIG_SAMV7_TC3_QE
+    case 3:
+      return &g_tc3lower;
+#endif
+    default:
+      return NULL;
+    }
+}
+
+/****************************************************************************
+ * Name: sam_setup
+ *
+ * Description:
+ *   This method is called when the driver is opened.  The lower half driver
+ *   should configure and initialize the device so that it is ready for use.
+ *   The initial position value should be zero.
+ *
+ ****************************************************************************/
+
+static int sam_setup(FAR struct qe_lowerhalf_s *lower)
+{
+  FAR struct sam_lowerhalf_s *priv = (FAR struct sam_lowerhalf_s *)lower;
+
+  /* Start the counter */
+
+  sam_tc_start(priv->tch);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: sam_shutdown
+ *
+ * Description:
+ *   This method is called when the driver is closed.  The lower half driver
+ *   should stop data collection, free any resources, disable timer hardware,
+ *   and put the system into the lowest possible power usage state
+ *
+ ****************************************************************************/
+
+static int sam_shutdown(FAR struct qe_lowerhalf_s *lower)
+{
+  FAR struct sam_lowerhalf_s *priv = (FAR struct sam_lowerhalf_s *)lower;
+
+  sam_tc_stop(priv->tch);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: sam_position
+ *
+ * Description:
+ *   Return the current position measurement.
+ *
+ ****************************************************************************/
+
+static int sam_position(FAR struct qe_lowerhalf_s *lower, FAR int32_t *pos)
+{
+  FAR struct sam_lowerhalf_s *priv = (FAR struct sam_lowerhalf_s *)lower;
+
+  /* Return the counter value */
+
+  *pos = (int32_t)(int16_t)sam_tc_getcounter(priv->tch);
```

Comment:
The only question is should this be `*pos = (int32_t)sam_tc_getcounter(priv->tch);` or `*pos = (int32_t)(int16_t)sam_tc_getcounter(priv->tch);`?